drivers: add API for VIRTIO devices and add VIRTIO PCI driver

fkokosinski opened the pull request:
This PR adds the API for accessing VIRTIO devices, and adds a driver for a VIRTIO PCI device based on the newly added API, and  adds a build test for the newly added VIRTIO PCI device. This PR is first of many, and more VIRTIO-related PRs will be submitted when this PR gets merged.

This PR is based on the Virtual I/O Device (VIRTIO) Version 1.3 specification: https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.pdf.

jmichalski-ant committed:
virtio: add API for VIRTIO devices and add VIRTIO PCI driver

This commit adds the API for accessing VIRTIO devices, and adds a driver
for a VIRTIO PCIE device based on the newly added API.

This commit is based on the Virtual I/O Device (VIRTIO) Version 1.3
specification:
https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.pdf

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

jmichalski-ant committed:
tests: add VIRTIO PCI build test

This commit adds a build test for the VIRTIO PCI device.

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

henrikbrixandersen commented:
@fkokosinski Will you be available for presenting this addition of a new device class to the Architecture WG? /CC @carlescufi

fkokosinski commented:
Hey @henrikbrixandersen, would February 11th work for you guys?

carlescufi commented:
> Hey @henrikbrixandersen, would February 11th work for you guys?

Yes, that would work.

cfriedt commented on drivers/virtio/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_VIRTIO_PCI virtio_pci.c virtqueue.c)
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_VIRTIO virtqueue.c)
zephyr_library_sources_ifdef(CONFIG_VIRTIO_PCI virtio_pci.c)
```

fkokosinski replied:
Done

cfriedt commented on tests/drivers/build_all/virtio/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+    - virtio
+tests:
+  drivers.virtio_pci.build:
+    platform_allow: qemu_x86_64
```
```suggestion
    filter: 
```

fkokosinski replied:
Done

cfriedt commented on tests/drivers/build_all/virtio/prj.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_TEST=y
+CONFIG_VIRTIO=y
+CONFIG_PCIE=y
+CONFIG_VIRTIO_PCI=y
```
defaults to `y` when DT node has `status = "okay"`
```suggestion
```

fkokosinski replied:
Done

cfriedt commented on drivers/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,668 @@
+};
+
+#define VIRTIO_PCI_DEFINE(inst)                                                     \
+	DEVICE_PCIE_INST_DECLARE(inst);                                                 \
```
Above this it might be good to ensure that the DT parent has compat `pcie-controller` with a `BUILD_ASSERT()`

fkokosinski replied:
Done

cfriedt reviewed (commented):
Looks pretty great to me other than some minor refinements.

Would be good for @ycsin or @luchnikov to take a look too (at fence operations, among other things).

carlescufi commented:
Architecture WG:

- @fkokosinski introduces the concept of VirtIO, and how this PR is the foundation of a framework in Zephyr that will be able to build upon in later PRs
- @nashif asks whether this PR can actually be tested at all. As per the author, this is not the case. Today this PR only introduces the building blocks, but it doesn't actually give you any functionality
- @nashif asks for an RFC where the roadmap towards the next steps after merging this PR are, so that we know what direction we're going towards. @cfriedt agrees with that
- @fkokosinski and @PiotrZierhoffer will provide an RFC issue

cfriedt commented on tests/drivers/build_all/virtio/testcase.yaml:
```diff
@@ -0,0 +1,12 @@
+tests:
+  drivers.virtio_pci.build: {}
+  drivers.virtio_pci.mmu.build:
+    filter: CONFIG_CPU_HAS_MMU
+    extra_configs:
+      - CONFIG_MMU=y
```
`MMU` is non longer a user-configurable option and can be thought of as semantically equivalent to `CONFIG_CPU_USE_MMU`. E.g. just because a CPU has an MMU does not mean the platform has configured and enabled the MMU for use.

```suggestion
    filter: CONFIG_MMU
```

fkokosinski commented:
>     * @nashif asks whether this PR can actually be tested at all. As per the author, this is not the case. Today this PR only introduces the building blocks, but it doesn't actually give you any functionality

Draft PR that uses the VIRTIO PCI driver can be found here: https://github.com/zephyrproject-rtos/zephyr/pull/86768

>     * @nashif asks for an RFC where the roadmap towards the next steps after merging this PR are, so that we know what direction we're going towards. @cfriedt agrees with that
>     * @fkokosinski and @PiotrZierhoffer will provide an RFC issue

RFC issue can be found here: https://github.com/zephyrproject-rtos/zephyr/issues/86561

FYI @cfriedt, @carlescufi

jmichalski-ant committed:
docs: virtio: add virtio docs

This commit adds virtio section in documentation, describing
virtio-related concepts

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

peter-mitsis commented on drivers/virtio/virtqueue.c:
```diff
@@ -0,0 +1,154 @@
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&v->mtx, K_FOREVER);
```
Is _virtq_add_buffer_chain()_ intended to ever be called from an ISR? As it is written, it must not be called from an ISR due to both the use of _K_FOREVER_ as a timeout parameter and the use of _k_mutex_lock()_. If not, then a comment in the routine's description stating that restriction.

fkokosinski replied:
Thanks for pointing this out. While the only device we implemented so far (virtiofs) doesn't need it, we've analyzed the other virtio devices and some of them may benefit from/require placing descriptors back in the available ring in the ISR directly after receiving them (mainly the devices with dedicated rx queue, where we would like to keep the available ring full and would possibly have no other place to call `virtq_add_buffer_chain`).

With the above in mind we modified this function to take a timeout param and changed mutex to a spinlock, to allow calling it from ISR with `K_NO_WAIT`.

cfriedt commented:
This did go through the arch review already several weeks ago. I wonder if it safe to remove the label @carlescufi ?

I've gone over it and with the addition of the extensive documentation, it looks good from my perspective.

@kartben might also like to review.

soburi commented on drivers/virtio/virtqueue.c:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	/* According to the spec 2.7.5.2 the maximum size of descriptor chain is 4GB */
+	if (total_len > GB(4)) {
```
In my testing, it seems that this `GB(4)` is rounded up
in qemu_x86 and cannot be replaced with the correct value.
You could fix util.h, but here you could work around it by defining the values ​​individually.

jmichalski-ant replied:
Done

soburi commented on tests/drivers/build_all/virtio/boards/qemu_x86_64.overlay:
```diff
@@ -0,0 +1,19 @@
+/*
```
I think it would be better to make it compatible with qemu_x86 as well.

jmichalski-ant replied:
Done

soburi reviewed (commented):
I created an entropy driver in #89439 and it works fine. However, I found some issues when compiling with qemu_x86, so I think it would be better to address them if possible.

soburi commented on drivers/CMakeLists.txt:
```diff
@@ -89,6 +89,7 @@ add_subdirectory_ifdef(CONFIG_SYSCON syscon)
 add_subdirectory_ifdef(CONFIG_SYS_CLOCK_EXISTS timer)
 add_subdirectory_ifdef(CONFIG_TEE tee)
 add_subdirectory_ifdef(CONFIG_VIDEO video)
+add_subdirectory_ifdef(CONFIG_VIRTIO_PCI virtio)
```
This line should be changed to
```
add_subdirectory_ifdef(CONFIG_VIRTIO virtio)
```

jmichalski-ant replied:
Done

soburi commented on include/zephyr/virtio/virtio.h:
```diff
@@ -0,0 +1,182 @@
+ * @param bit selected bit
+ * @return value of the selected bit
+ */
+static inline bool virtio_read_driver_feature_bit(const struct device *dev, int bit)
```
As feedback from the result of implementing virtio-mmio in #89460.

The driver features register is

https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.pdf#subsection.4.2.2

```
DriverFeatures 0x020 W
```

and cannot be read. If you are going to abstract virtio-mmio, it may be better not to have it.

soburi commented on drivers/virtio/virtio_pci.c:
```diff
@@ -0,0 +1,688 @@
+	const struct device *dev, virtio_enumerate_queues cb, void *opaque)
+{
+	struct virtio_pci_data *data = dev->data;
+	uint16_t queue_count = sys_le16_to_cpu(data->common_cfg->num_queues);
```
As feedback from the result of implementing virtio-mmio in https://github.com/zephyrproject-rtos/zephyr/pull/89460.

The virtio-mmio does not have a register equivalent to `num_queues`, so it is not possible to determine the upper limit of the number of queues.

In the Linux API `find_vqs`, the number of queues is explicitly specified from the driver side.

In the same way, it is recommended that an argument be added to the virtio_init_virtqueues API.

soburi commented on drivers/virtio/virtqueue.c:
```diff
@@ -0,0 +1,186 @@
+/*
```
BTW, BTW, can't we use open-amp's virtqueue implementations?

https://github.com/zephyrproject-rtos/open-amp/blob/main/open-amp/lib/virtio/virtqueue.c

fkokosinski replied:
Initially, we evaluated the use of the OpenAMP library, but we found that it supports only legacy VIRTIO. Meaning, among other things, its API limits us to the first 32 feature bits only, and there are no endianness conversions mandated by the specification.

cfriedt replied:
Good call to future-proof it. It would be ideal if OpenAMP could uprev at some point. I know that's a popular framework.

soburi replied:
@fkokosinski 
Thank you for checking. I fully understand the reason.

soburi reviewed (dismissed):
LGTM

fkokosinski commented:
@soburi we've separated ISR handling into `virtio_common`, so that stuff between PCI and MMIO can be shared (code-wise).

cfriedt commented:
@carlescufi - is there any additional need to keep this In Progress in Arch Review, or can it be moved to Done?

cfriedt commented:
@nashif @MaureenHelm - have you had a chance to review?
